### ISSUE TITLE
Qustion: How to Use Custom loss function with groupby optimizing error

### ISSUE THREAD

**Issue by innjoshka** [catboost/catboost#1560]
Hey, I use catboost for my ML task and I need to minimize error not for all data rows passed but for groups of rows. Each group for me have importance even if in 100k rows only 3 of them are under specific group and I want to minimize not GLOBAL error (because the group with 3 rows will have low importance on weights configuring if other group will have 80k rows but by groups and them sum total error up (so much preferable to minimize total error and catboost do it). I want Kinda create custom RMSE based on groupby minimizing where 3rows and 80k rows for different groups will affect the same on weights configuring. 

Let's look at some Rmse custom loss function.

```
class RmseObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        print(approxes, targets, weights)
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        result = []
        for index in range(len(targets)):
            der1 = targets[index] - approxes[index]
            der2 = -1

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))
        return result


model = CatBoostRegressor(loss_function=RmseObjective())
```

How can I modify it to be able to be able minimize by groups. 

By the way, approxes as I understand is already - y_pred and targets - y_actual am I right? If yes, then are there my data rows inside custom loss and how can i refer to it? 

Also as one more workaround to have different weights for each group in feature column but not create 100500 catboost models, because I have around 1000+ groups and amount constantly growth. Is it possible to do it using your catboost library?

Thanks for any help & suggestions

**Comment by Dan-Treacher** [comment]
Apologies if I'm not allowed to reply to these threads (I don't work for catboost), but for your first problem, if I understand you correctly, your problem is essentially that of regression with imbalanced classes in your predictive features (I have a similar problem).

Using the CatBoost Pool object (https://catboost.ai/docs/concepts/python-reference_pool.html), you can specify a `weight` parameter which can be applied either on a row by row basis, or on a group by group basis. I have a problem similar to yours, and I calculate weights for my groups based on inverse frequency (weight for group 'a' = 1/number of group 'a' members). If you then pass this column to the weights parameter in the pool object, they're applied automatically during the training.

I'm afraid I don't understand the last parts of your question

**Comment by innjoshka** [comment]
@Dan-Treacher Thanks for suggestion, could you provide example how to apply weights on group by group basis?. I have tried `group_weights` with `group_id` but nothing except errors I don't get. 

If I pass `weight` param, I have to pass weights for each row. As I understand correctly, I need to pass weight for each row based on inverse frequency and normalize by all rows suc as ((weight for group 'a' = 1/number of group 'a' members) / total_n_rows). Am I doing right?

**Comment by Dan-Treacher** [comment]
`# Import some libraries
import numpy as np
import pandas as pd
from catboost import CatBoostRegressor, Pool
from sklearn.model_selection import train_test_split


# Make up some data (In my application I have a mix of numeric and categorical data)
example_numerical_data = np.random.randn(20)*10
example_categorical_data = ['a']*5 + ['b']*10 + ['c']*2 + ['d']*3
example_targetVariable = np.random.randn(20)*10
df = pd.DataFrame({'feature1': example_numerical_data
                  , 'feature2': example_categorical_data
                  , 'targetVariable': example_targetVariable})
df['feature1'] = pd.to_numeric(df['feature1'])
df['feature2'] = df['feature2'].astype('category')
df['targetVariable'] = pd.to_numeric(df['targetVariable'])


# Calculate weights and add them as an extra column to df
frequencies = df['feature2'].value_counts()
df['weights'] = 0  # Initialise column
for cat in df['feature2']:
    df.loc[df['feature2'] == cat, 'weights'] = 1/frequencies.loc[cat]


# Split the testing and training data and put them into Pools with weights
X_train, X_test, y_train, y_test =train_test_split(df.drop(['targetVariable'], axis=1), df['targetVariable'])
training_data = Pool(data=X_train.drop('weights', axis=1)
                     , label=y_train
                     , cat_features=X_train.select_dtypes('category')
                     , weight=X_train['weights'])
testing_data = Pool(data=X_test.drop('weights', axis=1)
                    , label=y_test
                    , cat_features=X_test.select_dtypes('category')
                    , weight=None)  # Generally, I don't think you need to apply weights during testing because the model will already have learnt the weights during training...


# Do the regression and predictions
model = CatBoostRegressor(iterations=10)
model.fit(training_data)
predictions = model.predict(testing_data)
predictions`

**Comment by innjoshka** [comment]
@Dan-Treacher Thanks a lot for your help and time consumed on reply. It helped a lot. I have prototyped it with my data and got improvement. It occurred easier than I thought with weights configuring. Also It gives me ability to tune weights for specific groups to improve their affection on model predicts
